File: pset1/credit/credit.c
#include <cs50.h>
#include <stdio.h>

int return_digits(int n);

int main(void)
{

    long long int n = get_long_long("Number: ");
    //eprintf("n: %lld\n", n);
    int digit;
    int prev_digit;
    int doubled_digit;
    int loop_count = 0;
    int checksum = 0;

    while (n)
    {
        loop_count += 1;
        prev_digit = digit;
        digit = n % 10;
        n /= 10;
        //eprintf("loop: %i\n", loop_count);
        //eprintf("digit: %i\n", digit);

        if (loop_count % 2 == 0)
        {
            doubled_digit = digit * 2;
            if (doubled_digit >= 10)
            {
                //eprintf("doubled_digit: %i\n", doubled_digit);
                //if 10 or more add first then second digit
                //can't get a three digit number from add two one digits
                //so no need for a loop
                checksum += doubled_digit % 10;
                doubled_digit /= 10;
                checksum += doubled_digit;
            }
            else
            {
                checksum += doubled_digit;
            }
        }
        else
        {
            checksum += digit;
        }

        //eprintf("checksum = %i\n", checksum);

    }

    //default is number isn't valid
    string s = "INVALID";

    //test if checksum result valid of not
    if (checksum % 10 == 0)
    {
        //loop_count conditions for how many digits should be in card number
        if (digit == 4 && (loop_count == 13 || loop_count == 16))
        {
            s = "VISA";
        }
        else if (digit == 3 && (prev_digit == 4 || prev_digit == 7) && loop_count == 15)
        {
            s = "AMEX";
        }
        else if (digit == 5 && prev_digit >= 1 && prev_digit <= 5 && loop_count == 16)
        {
            s = "MASTERCARD";
        }

        //used to when checking if it works, but question asks for
        //only VISA, AMEX, MASTERCARD or INVALID to be the output
        //so commenting it out
        //else {s = "valid but diff card";}
    }

    printf("%s\n", s);
}

File: pset6/credit/credit.py
from cs50 import get_int

card = str(get_int("Number: "))
reversed_card = card[::-1]
check_sum = 0

for i in range(len(reversed_card)):
    digit = int(reversed_card[i])

    if i % 2 == 0:
        check_sum += digit
    else:
        doubled_digit = digit * 2
        if doubled_digit >= 10:
            first_digit = doubled_digit % 10
            second_digit = (doubled_digit - first_digit) / 10
            check_sum += first_digit + second_digit
        else:
            check_sum += doubled_digit

# Default case
card_provider = 'INVALID'

if check_sum % 10 == 0:
    if card[0] == '4' and (len(card) == 13 or len(card) == 16):
        card_provider = 'VISA'
    elif (card[0:2] == '34' or card[0:2] == '37') and len(card) == 15:
        card_provider = 'AMEX'
    elif int(card[0:2]) >= 51 and int(card[0:2]) <= 55 and len(card) == 16:
        card_provider = 'MASTERCARD'

print(card_provider)
